osrf/mbari_wec_gz
# Add mooring

Addresses #15 

Opening for visibility. Test isn't complete yet.

Adds mooring line, chain, and seafloor.

![2022-11-12-013357_2000x1690_scrot_chain](https://user-images.githubusercontent.com/7608908/203235552-bcdc7035-d277-4f06-bdb2-5b87de6094a6.png)

![2022-11-12-015245_2000x1690_scrot_chainJoints](https://user-images.githubusercontent.com/7608908/203235598-de0e0a65-3663-455d-a815-8ecd3d495b68.png)

https://user-images.githubusercontent.com/7608908/203235648-7375edbe-d450-42ce-be26-ccc1e6515851.mp4

## Test WIP

The test isn't able to push the buoy all the way to the bounds to test the bounds yet. Because the heave cone is so heavy, a very large force has to be applied to move the buoy just a few meters. After some iterations, an out of bounds error happens, seeming to do with collision, and the pose values go haywire.

3000 iterations of 10,000 N (is that the unit?) in x:
```
[Dbg] [mooring_drift.cpp:106] iter 2275, buoy link pose: 11.9083 0.003313 -2.44625
[Dbg] [mooring_drift.cpp:106] iter 2276, buoy link pose: -2.64042e+39 -9.35589e+38 6.45126e+38
/home/developer/mbari/buoy_sim_ws/src/buoy_sim/buoy_tests/tests/mooring_drift.cpp:108: Failure
Expected: ((buoyXY - anchorXY).Length()) < (maxRadius), actual: 2.80127e+39 vs 60

ODE INTERNAL ERROR 1: assertion "aabbBound >= dMinIntExact && aabbBound < dMaxIntExact" failed in collide() [collision_space.cpp:460]
Aborted (core dumped)
```

Note how the pose jumps from (12, 0, -2.4) to something huge. That's not the actual test failing (looks for < 60 m from anchor), that's values going out of bounds.

There are a few things I can try:
- Apply ocean currents (https://github.com/gazebosim/gz-sim/pull/800) manually on the command line and see how far that pushes the buoy
- Enable GUI in the test - is there a way to do that? I didn't find a GUI parameter in `ServerConfig` that we can pass to the server.
- Write a standalone plugin (basically identical to the one in the test) to run in the simulation and watch the GUI to check things are moving correctly.

My main uncertainty is that, since I'm not seeing the GUI, even before it goes out of bounds, I don't know if the buoy is moving to the force correctly.
